x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/rules_table/rules_table_context.tsx

```diff
@@ -210,43 +215,6 @@ export const RulesTableContextProvider = ({
         .slice((page - 1) * perPage, page * perPage)
     : filteredRules;
 
-  const queryClient = useQueryClient();
-  /**
-   * Use this method to update rules data cached by react-query.
-   * It is useful when we receive new rules back from a mutation query (bulk edit, etc.);
-   * we can merge those rules with the existing cache to avoid an extra roundtrip to re-fetch updated rules.
-   */
-  const updateRules = useCallback(
-    (newRules: Rule[]) => {
-      queryClient.setQueryData<ReturnType<typeof useFindRules>['data']>(
-        getFindRulesQueryKey({ isInMemorySorting, filterOptions, sortingOptions, pagination }),
-        (currentData) => ({
-          rules: mergeRules(currentData?.rules || [], newRules),
-          total: currentData?.total || 0,
-        })
-      );
-
-      /**
-       * Unset loading state for all new rules
-       */
-      const newRuleIds = newRules.map((r) => r.id);
-      const newLoadingRuleIds = loadingRules.ids.filter((id) => !newRuleIds.includes(id));
-      setLoadingRules({
-        ids: newLoadingRuleIds,
-        action: newLoadingRuleIds.length === 0 ? null : loadingRules.action,
-      });
-    },
```

Comment:
I think my only concern with this PR is that updating the cache of loaded rules has been replaced by cache invalidation. I'd expect that in the in-memory mode, this can lead to continuous reloading of all the rules when the user is actively working with the table (e.g. enabling/disabling rules, editing tags, etc).

What was the reason for this change?

If the reason is that now we have two caches that are maintained simultaneously (`allRules` and `pagedRules` in `useFindRules`), could we change the implementation of `updateRules` so that it updates both of them?

Reply:
> I'd expect that in the in-memory mode, this can lead to continuous reloading of all the rules when the user is actively working with the table (e.g. enabling/disabling rules, editing tags, etc).

This change does not affect editing operations because they never had optimistic updates. Therefore, only rule (de)activation is affected. So while cache invalidation is not an ideal option, I don't think it will significantly impact the table's performance.

> If the reason is that now we have two caches that are maintained simultaneously (allRules and pagedRules in useFindRules), could we change the implementation of updateRules so that it updates both of them?

Yes, we have already discussed it on yesterday's sync. A more intelligent cache update algorithm would be a better option. But given the timeframe before FF, there's no chance I can implement it in this PR.

Reply:
Synced on that with @xcrzx in more detail, the agreement is that in general, it makes sense to update the cache instead of invalidating it (when we enable/disable a rule, update it, etc), but

1) we never had it in the implementation of the table previously for most of the rule actions, except the enable/disable action where we have optimistic updates
2) it's too late to change that in this PR

Also, the agreement is that updating the rules cache will go as a follow-up performance optimization:

- Implement optimistic updates instead of cache invalidation for rule activation/deactivation
- Implement optimistic updates instead of cache invalidation for bulk operations (need support on the API side)